Move rounds related logic from account.merge function to rounds module - Closes #2942

SargeKhan opened the pull request:
### What was the problem?
The account.merge function should only update data relevant to accounts. The entries added to rounds tables should be handled in the rounds module.
### How did I fix it?
Moved database queries related to rounds to the rounds module.
### How to test it?
Run unit tests.
### Review checklist

* The PR resolves #2942 
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

shuse2 commented on framework/src/modules/chain/logic/account.js:
```diff
@@ -446,16 +445,6 @@ class Account {
 							);
 						}
 
-						if (updatedField === 'balance') {
```
can we already remove this? I thought `reward` part still goes through this path

SargeKhan replied:
You are right. :( I will add this logic again. But this means that as long as we don't remove the transactions implemented in the core, we will have a different delegates list.

SargeKhan commented on framework/test/unit/modules/chain/unit/modules/rounds.js:
```diff
@@ -237,6 +246,282 @@ describe('rounds', async () => {
 		});
 	});
 
+	describe('__private.updateRecipientsRoundInformationWithAmountForTransactions', () => {
```
We don't have this convention in core.

shuse2 commented on framework/test/unit/modules/chain/unit/modules/rounds.js:
```diff
@@ -237,6 +246,282 @@ describe('rounds', async () => {
+		});
+
+		it('should call createRoundInformationWithAmount with negative value when forwardTick is set to true', async () => {
+			updateSendersRoundInformationWithAmountForTransactions(
```
should we `await` here?
same for below

nazarhussain commented on framework/src/modules/chain/modules/rounds.js:
```diff
@@ -62,6 +64,128 @@ class Rounds {
+		.then(() => setImmediate(cb))
+		.catch(err => setImmediate(cb, err));
+};
+
```
Add js-doc to new created functions.

nazarhussain commented on framework/src/modules/chain/modules/rounds.js:
```diff
@@ -62,6 +64,128 @@ class Rounds {
 	}
 }
 
+__private.updateSendersRoundInformationWithAmountForTransactions = function(
```
`updateSendersRoundInformationWithAmount` is good enough to explain. We may ned up function names with line character limit.

nazarhussain commented on framework/src/modules/chain/modules/rounds.js:
```diff
@@ -62,6 +64,128 @@ class Rounds {
+	// For IN_TRANSFER TRANSACTION, we first need to get the authorId to get the recipient
+	return Promise.all(
+		transactions
+			.filter(transaction => transaction.type === transactionTypes.IN_TRANSFER)
```
`filter( ({ transaction: type }) => type === transactionTypes.IN_TRANSFER)`

nazarhussain commented on framework/src/modules/chain/modules/rounds.js:
```diff
@@ -62,6 +64,128 @@ class Rounds {
+				// Check the owner of the dapp for recipient account
+				library.storage.entities.Transaction.getOne(
+					{
+						id: transaction.asset.inTransfer.dappId,
```
You can first map to get the ids of the transactions and then use `id_in` to avoid running multiple queries against active transaction connection.

nazarhussain commented on framework/src/modules/chain/modules/rounds.js:
```diff
@@ -62,6 +64,128 @@ class Rounds {
+		.map(transaction =>
+			(forwardTick
+				? transaction.asset.votes
+				: Diff.reverse(transaction.asset.votes)
```
We should use builtin functions for such basic behaviours. Having an helper for just for `Array.reverse` is not a good approach.

shuse2 replied:
this `reverse` is changing the sign of the `vote`, not reversing the transaction

nazarhussain commented on framework/src/modules/chain/modules/rounds.js:
```diff
@@ -62,6 +64,128 @@ class Rounds {
+			).map(vote => {
+				// Fetch first character
+				const mode = vote[0];
+				const dependentId = vote.slice(1);
```
This was the old logic to get `dependentId`, its not safe to replace with simply `slice`.

```
let mode = updatedValueItem[0];
let dependentId = '';

if (mode === '-' || mode === '+') {
	dependentId = updatedValueItem.slice(1);
} else {
	dependentId = updatedValueItem;
	mode = '+';
}
```

shuse2 replied:
At this point, every transaction has to be valid in all the case. Therefore, it should be safe to do this.
Also, it's kinda scary to update the `mode` if it's not + or -....

nazarhussain commented on framework/src/modules/chain/modules/rounds.js:
```diff
@@ -62,6 +64,128 @@ class Rounds {
+			forwardTick,
+			tx
+		),
+	].reduce((accPromises, currPromises) => accPromises.concat(currPromises), []);
```
Why we even need to use `reduce` here?

shuse2 replied:
each function returns Array<Promise>, so this reduce is same as `flatmap`

nazarhussain commented on framework/src/modules/chain/modules/rounds.js:
```diff
@@ -247,12 +379,20 @@ Rounds.prototype.tick = function(block, done, tx) {
+				__private.updateRoundInformationForTransactions(
+					round,
+					block.transactions,
+					true,
```
In recent code, we are using a constant approach rather than passing static values. e.g. 

```
const isForwardTick = true; 
```

And then pass `isForwardTick` to function, that means code more readable.

SargeKhan commented:
I merged this PR because it already had 2 reviews and I was open for quite some time. But thank you for your reviews. I will create a separate pull request addressing these comments.
